File: mobile/src/main/java/us/jbmadscience/fractalrush/Mandelbrot.java
package com.example.android.opengl;

import android.opengl.GLES20;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.FloatBuffer;
import java.nio.ShortBuffer;

/**
 * Created by Joshua on 2/21/2016.
 */
public class Mandelbrot {
    private final FloatBuffer vertexBuffer;
    private final ShortBuffer drawListBuffer;
    private final int mProgram;
    private int mPositionHandle;
    private int mMVPMatrixHandle;

    // number of coordinates per vertex in this array
    static final int DIMENSIONS = 3;

    static float viewVertices[] = {
            -1.0f,  1.0f, 0.0f,   // top left
            -1.0f, -1.0f, 0.0f,   // bottom left
            1.0f, -1.0f, 0.0f,   // bottom right
            1.0f,  1.0f, 0.0f }; // top right

    private final short drawOrder[] = { 0, 1, 2, 0, 2, 3 }; // order to draw vertices
    private final int vertexRecordSize = DIMENSIONS * 4; // 4 bytes per vertex

    public Mandelbrot(String viewVertexShaderCode, String mandelbrotShaderCode) {
        // initialize vertex byte buffer for shape coordinates
        ByteBuffer bb = ByteBuffer.allocateDirect(
                // (# of coordinate values * 4 bytes per float)
                viewVertices.length * 4);
        bb.order(ByteOrder.nativeOrder());
        vertexBuffer = bb.asFloatBuffer();
        vertexBuffer.put(viewVertices);
        vertexBuffer.position(0);

        // initialize byte buffer for the draw list
        ByteBuffer dlb = ByteBuffer.allocateDirect(
                // (# of coordinate values * 2 bytes per short)
                drawOrder.length * 2);
        dlb.order(ByteOrder.nativeOrder());
        drawListBuffer = dlb.asShortBuffer();
        drawListBuffer.put(drawOrder);
        drawListBuffer.position(0);

        // Prepare shaders
        int vertexShader = MandelbrotRenderer.loadShader(
                GLES20.GL_VERTEX_SHADER,
                viewVertexShaderCode);
        int fragmentShader = MandelbrotRenderer.loadShader(
                GLES20.GL_FRAGMENT_SHADER,
                mandelbrotShaderCode);

        // create empty OpenGL Program
        mProgram = GLES20.glCreateProgram();

        // add the vertex shader to program
        GLES20.glAttachShader(mProgram, vertexShader);
        // add the fragment shader to program
        GLES20.glAttachShader(mProgram, fragmentShader);
        // create OpenGL program executables
        GLES20.glLinkProgram(mProgram);
    }

    /**
     * Encapsulates the OpenGL ES instructions for drawing this shape.
     *
     * @param mvpMatrix - The Model View Project matrix in which to draw
     * this shape.
     */
    public void draw(float[] mvpMatrix) {
        // Add program to OpenGL environment
        GLES20.glUseProgram(mProgram);

        // get handle to vertex shader's vPosition member
        mPositionHandle = GLES20.glGetAttribLocation(mProgram, "vPosition");
        mMVPMatrixHandle = GLES20.glGetUniformLocation(mProgram, "uMVPMatrix");

        GLES20.glUniformMatrix4fv(mMVPMatrixHandle, 1, false, mvpMatrix, 0);

        GLES20.glEnableVertexAttribArray(mPositionHandle);
        GLES20.glVertexAttribPointer(
                mPositionHandle, DIMENSIONS,
                GLES20.GL_FLOAT, false,
                vertexRecordSize, vertexBuffer);

        // Draw the square
        GLES20.glDrawElements(
                GLES20.GL_TRIANGLES, drawOrder.length,
                GLES20.GL_UNSIGNED_SHORT, drawListBuffer);

        // Disable vertex array
        GLES20.glDisableVertexAttribArray(mPositionHandle);
    }


}

File: mobile/src/main/java/us/jbmadscience/fractalrush/Mandelbrot.kt
package us.jbmadscience.fractalrush

import android.opengl.GLES20

import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.FloatBuffer
import java.nio.ShortBuffer

/**
 * Created by Joshua on 2/21/2016.
 */
class Mandelbrot(viewVertexShaderCode: String, mandelbrotShaderCode: String) {
    private val _vertexBuffer: FloatBuffer
    private val _drawListBuffer: ShortBuffer
    private val _program: Int
    private var _positionHandle: Int = 0
    private var _mvpMatrixHandle: Int = 0

    private val drawOrder = shortArrayOf(0, 1, 2, 0, 2, 3) // order to draw vertices
    private val vertexRecordSize = DIMENSIONS * 4 // 4 bytes per vertex

    init {
        // initialize vertex byte buffer for shape coordinates
        val bb = ByteBuffer.allocateDirect(
                // (# of coordinate values * 4 bytes per float)
                viewVertices.size * 4)
        bb.order(ByteOrder.nativeOrder())
        _vertexBuffer = bb.asFloatBuffer()
        _vertexBuffer.put(viewVertices)
        _vertexBuffer.position(0)

        // initialize byte buffer for the draw list
        val dlb = ByteBuffer.allocateDirect(
                // (# of coordinate values * 2 bytes per short)
                drawOrder.size * 2)
        dlb.order(ByteOrder.nativeOrder())
        _drawListBuffer = dlb.asShortBuffer()
        _drawListBuffer.put(drawOrder)
        _drawListBuffer.position(0)

        // Prepare shaders
        val vertexShader = FractalRushRenderer.loadShader(
                GLES20.GL_VERTEX_SHADER,
                viewVertexShaderCode)
        val fragmentShader = FractalRushRenderer.loadShader(
                GLES20.GL_FRAGMENT_SHADER,
                mandelbrotShaderCode)

        _program = GLES20.glCreateProgram()

        GLES20.glAttachShader(_program, vertexShader)
        GLES20.glAttachShader(_program, fragmentShader)
        GLES20.glLinkProgram(_program)
    }

    fun draw(mvpMatrix: FloatArray) {
        // Add program to OpenGL environment
        GLES20.glUseProgram(_program)

        // get handle to vertex shader's vPosition member
        _positionHandle = GLES20.glGetAttribLocation(_program, "vPosition")
        _mvpMatrixHandle = GLES20.glGetUniformLocation(_program, "uMVPMatrix")

        GLES20.glUniformMatrix4fv(_mvpMatrixHandle, 1, false, mvpMatrix, 0)

        GLES20.glEnableVertexAttribArray(_positionHandle)
        GLES20.glVertexAttribPointer(
                _positionHandle, DIMENSIONS,
                GLES20.GL_FLOAT, false,
                vertexRecordSize, _vertexBuffer)

        // Draw the square
        GLES20.glDrawElements(
                GLES20.GL_TRIANGLES, drawOrder.size,
                GLES20.GL_UNSIGNED_SHORT, _drawListBuffer)

        // Disable vertex array
        GLES20.glDisableVertexAttribArray(_positionHandle)
    }

    companion object {

        // number of coordinates per vertex in this array
        private val DIMENSIONS = 3

        private val viewVertices = floatArrayOf(-1.0f, 1.0f, 0.0f, // top left
                -1.0f, -1.0f, 0.0f, // bottom left
                1.0f, -1.0f, 0.0f, // bottom right
                1.0f, 1.0f, 0.0f) // top right
    }


}
